docs/man/spamilter.8
.TH SPAMILTER 8 "0.60" "May 30, 2012" ""
.SH NAME
.B Spamilter
\- A libmilter based spamfilter
.SH SYNOPSIS
spamilter [-d level] [-c config file path filename] [-f]
.br
.SH OPTIONS
.TP
.B -d
[numeric] - debug level, 1 or 2. NB. This implies -f
.TP
.B -c
[string] - config file path filename - default = /etc/spamilter.rc
.TP
.B -f
[boolean] - run spamilter in the foreground
.SH DESCRIPTION
Spamilter is a Sendmail milter written entirely in C, and therefore is faster and less cpu intensive than other interperative based solutions.
.br
It blocks spam using the following methods;
.RS
.IP \(bu 4
Configurable Realtime DNS Blocklists
.IP \(bu 4
Sender/Reply-To Address devilvery verification
.IP \(bu 4
Configurable Sender/Recipient Black and White lists
.IP \(bu 4
Invalid MTA hostname verification
.IP \(bu 4
Basic Virus/Worm file attachment rejection for files ending in .pif, .scr, etc.. via MsExtChk filter
.IP \(bu 4
SPF via libspf
.IP \(bu 4
GeoIP Country Code block listing
.RE
Also;
.RS
.IP \(bu 4
Realtime firewall blocking of MTA hosts with invalid host names via
.B ipfwmtad
.IP \(bu 4
And realtime rate limited connection blocking, also via
.B ipfwmtad
.IP \(bu 4
Greylisting via
.B greydbd
.RE
.br
All actions are logged via syslog with both the sender and the recipient.
From this, report generation and notification to recipients showing activity becomes extremely simple.
.br
.SH EXAMPLES
spamilter -d 1
.SH VERSION
This documentation describes
.B Spamilter
version 0.60
.SH "SEE ALSO"
ipfwmtad(8), greydbd(8)
.br
.I http://www.wanlink.com/spamilter/
.SH AUTHOR
.br
Neal Horman \<spamilter@wanlink.com\>


Changelog
Changelog;
	20150129
	* mod - be more aggresive about ipfw blocking MTAs when in blacklist type of conditions
	* fix - don't double ipfw inculpate ip address

	20150125
	* mod - configure - do some clean up / modernization
	* fix - segfault in badext.c when logging 'unable to open file'
	* mod - change 'configure --sendmail-dir' can work on FreeBSD
	* mod - change 'configure --with-libspf' back to previous behavior, and split builtin behavior out to '--with-libspf-builtin'
	* add - 'make install-config' target to do initial config file installation
	* add - support for regex pattern matching for BWLIST mbox

	20150124
	* mod - mlfi_regex_mboxsplit() add '=' to valid characters list of mbox regex pattern
	* mod - fix dnsbl regression

	20150123
	* add - dns, dbl and dnsbl changes to grok ipv6

	20150121
	* mod - add ipv6 lookups for MtaHostChk functionality

	* tag - build_140227
	* mod - un-nest callback functions for llvm compatiblity

	* tag - build_130713_1
	* add - HLO host DBL checking
	* add - body host DBL checking

	* tag - build_121118_1
	* add - session id for logging purposes

	* tag - build_121006
	* add - do geoip checking for 'X-Originating-IP' hotmail header

	* fix - don't ignore SMFIS result from BodyHost checking function
	* mod - tweak replyto BWL handling behavior

	* mod - don't do Body URL black listing and Body URL CC black listing if there is rcpt or sndr Accept
		action, or is localnet host
	* mod - reply-to header delivery checking to ignore '<>' return path
	* add - setproctitle for "watcher" and "worker"

	* tag - build_120625_2
	* fix - regression of dns_query_rr_a() function use by "Invalid MTA hostname" filter

	* tag - build_120625_1
	* mod - bwlist to support postgresql
	* mod - table driver framework to support postgresql table driver
	* add - add postgresql table driver

	* tag - build_120618_1
	* mod - rewrite bwlist to be based off of the table framework and the flat file table driver
	* add - add flat file table driver
	* add - generic table driver framework

	* add - filter - using regex, scan the body for http/https/ftp urls and do BWL and geoip contry code lookup rejection

	* add - filter - received by/from header hostname BWL checking and geoip contry code ip lookup rejection
	* mod - change all injected headers to use the basic format of "X-Milter: Spamilter DataSet=xxx; foo=bar; foo=bar; ..."
	* add - filter - BWL and geoip country code lookup rejection for connecting mta

	* mod - start working towards ipv6 support

	* add - filter - reply to header return address verification
	* mod - test all RBL entries for the connecting mta ip address, to inject headers when tagging instead of rejecting

	* add - filter - grey listing mechanism (requires postgresl)
	* add - filter - fwdhost lookup - allow an edge MTA to consult an athorative inner MTA for recipent delivery status

	* add - ability for local definitions of localnet testing in ifi_islocalnet() via addtions to the struct _localnethosts
		defintion from the ifilocal.inc file
	* add - framework for local content filtering via c code from hndlrs_contentfilter_local.inc file
	* add - framework for local message duplication based on sender or recipient via hard coded rules in c code from
		dupelocal.inc file
	* mod - treat SASL authorized connections as a localnet host

	* add - filter to reject if recipent is not a target listed in virtusertable.db
	* add - filter to reject if recipent is not a target listed in aliases.db
	* add - filter to reject if recipent is not a local system user
	* mod - don't do dns lookups on a given rbl host until the reject/tag stage as listed/configured in spamilter.rc
		instead of doing the lookups at connect and then reject/tagging at stage
	* fix - local asprintf implementation in nstring.c to not segfault on a %s parameter when NULL
	* fix - null pointer sanity tests
	* add - cli -f to spamilter to keep it in foreground with debug info
	* mod - to EOM handler to tag or flag with all filter results instead of just one result
	* fix - for sender address verification to treat the mbox in a case insensative way per RFC2821
	* mod - don't collect body if not used
	* fix - debugmode > 1 works now

version - 0.60 - 051127
	* fix - actually use the configured db path instead of the hard coded one - patch from Luns Tee
	* fix - make sure the priv->smtprc is initialized. - patch from Luns Tee
	* fix - remove NULL assignment of argptr, Linux 7.? barfs on it
	* fix - Makefile.tmpl so that SunOs CC can deal with += vars, add -lpthread for SunOs
	* fix - memory leak in ifi.c - patch from Luns Tee
	* fix - memory leak in dnsbl.c - patch from Luns Tee
	* mod - to combat rfc ignorant mta hosts, try twice to negotiate an accepted
		"mail from:, rcpt to:" sequence, first with <>, then with <postmaster@`local host name`>
		if that fails.
	* fix - globally use mx_res_search that resizes the query buffer as needed - drastically altered
		patch from Luns Tee
	* mod - if not doing attachment checks, don't do body processing - patch from Luns Tee
	* mod - add --with-???? options to configure. see configure --help for more details
	* mod - use res_n???? thread safe resolver library calls
	* mod - rename spf.[ch] to spfapi.[ch] so as to not conflict with libspf's spf.h
	* mod - drop the included libspf in favor of using the "current" version as published on the
		web site at http://www.libspf.org now that --with-libspf is available in configure
		ntbw. installation of the library is not required, rather just configuration and
		compilation, as it is statically linked.
	* mod - ipfwmtad to handle exculpate correctly
	* add - ipfwmtad -u ipfw rule number
	* add - initial support for OpenBSD
	* mod - move MsExtChk core code into separate source module
	* add - support for reading MsExtChk file list from db.extensions in lue of being hard coded
	* add - support for ipfw2 in ipfwmtad_direct
	* mod - compile with Wall by default

src/spamilter.c
/*--------------------------------------------------------------------*
 *
 * Developed by;
 *	Neal Horman - http://www.wanlink.com
 *	Copyright (c) 2003 Neal Horman. All Rights Reserved
 *
 *	Redistribution and use in source and binary forms, with or without
 *	modification, are permitted provided that the following conditions
 *	are met:
 *	1. Redistributions of source code must retain the above copyright
 *	   notice, this list of conditions and the following disclaimer.
 *	2. Redistributions in binary form must reproduce the above copyright
 *	   notice, this list of conditions and the following disclaimer in the
 *	   documentation and/or other materials provided with the distribution.
 *	3. All advertising materials mentioning features or use of this software
 *	   must display the following acknowledgement:
 *	This product includes software developed by Neal Horman.
 *	4. Neither the name Neal Horman nor the names of any contributors
 *	   may be used to endorse or promote products derived from this software
 *	   without specific prior written permission.
 *	
 *	THIS SOFTWARE IS PROVIDED BY NEAL HORMAN AND ANY CONTRIBUTORS ``AS IS'' AND
 *	ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *	IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 *	ARE DISCLAIMED.  IN NO EVENT SHALL NEAL HORMAN OR ANY CONTRIBUTORS BE LIABLE
 *	FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 *	DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 *	OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *	HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *	LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 *	OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 *	SUCH DAMAGE.
 *
 *	CVSID:  $Id: spamilter.c,v 1.54 2013/07/13 23:31:37 neal Exp $
 *
 * DESCRIPTION:
 *	application:	spamilter
 *	module:		spamilter.c
 *--------------------------------------------------------------------*/

static char const cvsid[] = "@(#)$Id: spamilter.c,v 1.54 2013/07/13 23:31:37 neal Exp $";

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <syslog.h>
#include <sysexits.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <pwd.h>

#include "spamilter.h"
#include "watcher.h"
#include "dnsbl.h"
#include "hndlrs.h"
#include "misc.h"

int	gDebug			= 0;
int	gForeground		= 0;
char	*gPolicyUrl		= "http://www.somedomain.com/policy.html";
char	*gDbpath		= "/var/db/spamilter";
char	*gConfpath		= "/etc/spamilter.rc";
char	*gMlficonn		= "inet:7726@localhost";
int	gDnsBlChk		= 1;
int	gSmtpSndrChk		= 1;
char	*gSmtpSndrChkAction	= "Reject";
int	gMtaHostChk		= 1;
int	gMtaHostChkAsIp		= 1;
int	gMtaHostIpfw		= 0;
int	gMtaHostIpfwNominate	= 0;
int	gMtaHostIpChk		= 0;
int	gMsExtChk		= 1;
char	*gMsExtChkAction	= "Reject";
char	gHostnameBuf[1024];	// hope this is big enough!
char	*gHostname		= gHostnameBuf;
char	*gUserName		= "nobody";
#ifdef SUPPORT_POPAUTH
char	*gPopAuthChk		= NULL;
#endif
#ifdef SUPPORT_LIBSPF
int	gMtaSpfChk		= 0;
#endif
#ifdef SUPPORT_VIRTUSER
char	*gVirtUserTableChk	= NULL;
#endif
#ifdef SUPPORT_ALIASES
char	*gAliasTableChk		= NULL;
#endif
#ifdef SUPPORT_LOCALUSER
int	gLocalUserTableChk	= 0;
#endif
#ifdef SUPPORT_GREYLIST
int gGreyListChk		= 0;
#endif
#ifdef SUPPORT_FWDHOSTCHK
int gRcptFwdHostChk		= 0;
#endif
int gHeaderChkReplyTo		= 0;
int gHeaderChkReceived		= 0;
#ifdef SUPPORT_GEOIP
char *gpGeoipDbPath		= NULL;
int gGeoIpCcChk			= 0;
#endif

struct smfiDesc mlfi =
{
	"Spamilter",	// filter name
	SMFI_VERSION,	// version code -- do not change
	SMFIF_ADDHDRS | SMFIF_CHGHDRS | SMFIF_CHGBODY | SMFIF_ADDRCPT | SMFIF_DELRCPT,	// flags
	mlfi_connect,	// connection info filter
	mlfi_helo,	// SMTP HELO command filter
	mlfi_envfrom,	// envelope sender filter
	mlfi_envrcpt,	// envelope recipient filter
	mlfi_header,	// header filter
	mlfi_eoh,	// end of header
	mlfi_body,	// body block filter
	mlfi_eom,	// end of message
	mlfi_abort,	// message aborted
	mlfi_close	// connection cleanup
};

#define SetKeyValStr(k,v,n,d) { if(strlen(k) && strlen(v) && strcasecmp((k),(n)) == 0) (d) = strdup((v)); }
#define SetKeyValInt(k,v,n,d) { if(strlen(k) && strlen(v) && strcasecmp((k),(n)) == 0) (d) = atoi((v)); }

#define ShowKeyValStr(k,v) { printf("%s = '%s'\n",(k),(v)); }
#define ShowKeyValInt(k,v) { printf("%s = %u\n",(k),(v)); }

void getconf(char *confpath)
{	FILE	*fin = fopen(confpath,"r");
	char	*str;
	char	buf[1024];
	char	key[1024];
	char	val[1024];

	if(fin != NULL)
	{
		while(!feof(fin))
		{
			fgets(buf,sizeof(buf),fin);
			
			str = strchr(buf,'#');
			if(str != NULL)
			{
				*(str--) = '\0';
				while(str >= buf && (*str ==' ' || *str == '\t'))
					*(str--) = '\0';
			}

			if(strlen(buf))
			{
				str = mlfi_strcpyadv(key,sizeof(key),buf,'=');
				str = mlfi_strcpyadv(val,sizeof(val),str,'=');

				SetKeyValStr(key,val,"UserName",gUserName);
				SetKeyValStr(key,val,"PolicyUrl",gPolicyUrl);
				SetKeyValStr(key,val,"DbPath",gDbpath);
				SetKeyValStr(key,val,"Conn",gMlficonn);
				SetKeyValInt(key,val,"DnsBlChk",gDnsBlChk);
				SetKeyValInt(key,val,"SmtpSndrChk",gSmtpSndrChk);
				SetKeyValStr(key,val,"SmtpSndrChkAction",gSmtpSndrChkAction);
				SetKeyValInt(key,val,"MtaHostChk",gMtaHostChk);
				SetKeyValInt(key,val,"MtaHostChkAsIp",gMtaHostChkAsIp);
				SetKeyValInt(key,val,"MtaHostIpfw",gMtaHostIpfw);
				SetKeyValInt(key,val,"MtaHostIpfwNominate",gMtaHostIpfwNominate);
				SetKeyValInt(key,val,"MtaHostIpChk",gMtaHostIpChk);
#ifdef SUPPORT_LIBSPF
				SetKeyValInt(key,val,"MtaSpfChk",gMtaSpfChk);
#endif

				SetKeyValInt(key,val,"MsExtChk",gMsExtChk);
				SetKeyValStr(key,val,"MsExtChkAction",gMsExtChkAction);
#ifdef SUPPORT_POPAUTH
				SetKeyValStr(key,val,"PopAuthChk",gPopAuthChk);
#endif
#ifdef SUPPORT_VIRTUSER
				SetKeyValStr(key,val,"VirtUserTableChk",gVirtUserTableChk);
#endif
#ifdef SUPPORT_ALIASES
				SetKeyValStr(key,val,"AliasTableChk",gAliasTableChk);
#endif
#ifdef SUPPORT_LOCALUSER
				SetKeyValInt(key,val,"LocalUserTableChk",gLocalUserTableChk);
#endif
#ifdef SUPPORT_GREYLIST
				SetKeyValInt(key,val,"GreyListChk",gGreyListChk);
#endif
#ifdef SUPPORT_FWDHOSTCHK
				SetKeyValInt(key,val,"RcptFwdHostChk",gRcptFwdHostChk);
#endif
				SetKeyValInt(key,val,"HeaderReplyToChk",gHeaderChkReplyTo);
				SetKeyValInt(key,val,"HeaderReceivedChk",gHeaderChkReceived);
#ifdef SUPPORT_GEOIP
				SetKeyValStr(key,val,"GeoIPDBPath",gpGeoipDbPath);
				SetKeyValInt(key,val,"GeoIPChk",gGeoIpCcChk);
#endif
			}
		}
		fclose(fin);
	}
	else
		printf("Warning! Unable to open config file '%s', using compiled defaults.\n",confpath);
}

void usage()
{
	printf("spamilter [-c config] [-d 1] \n");
	printf("\tWhere;\n\t-c specifies a config filename to read at startup\n");
	printf("\t-d debug mode.\n");
}

int worker_main()
{
	return smfi_main();
}

int main(int argc, char *argv[])
{	int		c;
	struct passwd	*pw = NULL;
	int		uid = getuid();
	int		flags = 0;

	// Process command line options
	while ((c = getopt(argc, argv, "fd:c:")) != -1)
	{
		switch (c)
		{
			case 'c':
				if (optarg != NULL && *optarg != '\0')
					gConfpath = optarg;
				break;
			case 'd': // debug
				if(optarg != NULL && *optarg)
					gDebug = atoi(optarg);
				break;
			case 'f': // forground
				gDebug = 0;
				gForeground = 1;
				break;
			default:
				usage();
				exit(0);
				break;
		}
	}

	setproctitle("startup");

	getconf(gConfpath);

	if(gDebug)
		flags = LOG_PERROR;
	openlog(mlfi.xxfi_name, flags|LOG_NDELAY|LOG_PID, LOG_DAEMON);

	if(uid == 0) // if root, drop privs
	{
		if((pw = getpwnam(gUserName)) == NULL)
		{
			fprintf(stderr,"Fatal error - Unable to get user '%s' identity information",gUserName);
			exit(1);
		}
		else if(setgid(pw->pw_gid) != 0 || setuid(pw->pw_uid) != 0)
		{
			fprintf(stderr,"Fatal error - Unable to switch user identity to '%s'",gUserName);
			exit(2);
		}
		else
			uid = pw->pw_uid;
	}
	pw = getpwuid(uid);

	printf("\nStarting %s\n",mlfi.xxfi_name);
	printf("Running as user '%s/%s'\n",pw->pw_name,pw->pw_gecos);
	ShowKeyValStr("PolicyUrl",gPolicyUrl);
	ShowKeyValStr("DbPath",gDbpath);
	ShowKeyValStr("Conn",gMlficonn);
	ShowKeyValInt("DnsBlChk",gDnsBlChk);
	ShowKeyValInt("SmtpSndrChk",gSmtpSndrChk);
	ShowKeyValStr("SmtpSndrChkAction",gSmtpSndrChkAction);
	ShowKeyValInt("MtaHostChk",gMtaHostChk);
	ShowKeyValInt("MtaHostIpfw",gMtaHostIpfw);
	ShowKeyValInt("MtaHostIpfwNominate",gMtaHostIpfwNominate);
	ShowKeyValInt("MtaHostIpChk",gMtaHostIpChk);
#ifdef SUPPORT_LIBSPF
	ShowKeyValInt("MtaSpfChk",gMtaSpfChk);
#endif
	ShowKeyValInt("MsExtChk",gMsExtChk);
	ShowKeyValStr("MsExtChkAction",gMsExtChkAction);
#ifdef SUPPORT_POPAUTH
	ShowKeyValStr("PopAuthChk",gPopAuthChk);
#endif
#ifdef SUPPORT_VIRTUSER
	ShowKeyValStr("VirtUserTableChk",gVirtUserTableChk);
#endif
#ifdef SUPPORT_ALIASES
	ShowKeyValStr("AliasTableChk",gAliasTableChk);
#endif
#ifdef SUPPORT_LOCALUSER
	ShowKeyValInt("LocalUserTableChk",gLocalUserTableChk);
#endif
#ifdef SUPPORT_GREYLIST
	ShowKeyValInt("GreyListChk",gGreyListChk);
#endif
#ifdef SUPPORT_FWDHOSTCHK
	ShowKeyValInt("RcptFwdHostChk",gRcptFwdHostChk);
#endif
	ShowKeyValInt("HeaderReplyToChk",gHeaderChkReplyTo);
	ShowKeyValInt("HeaderReceivedChk",gHeaderChkReceived);
#ifdef SUPPORT_GEOIP
	ShowKeyValStr("GeoIPDBPath",gpGeoipDbPath);
	ShowKeyValInt("GeoIPChk",gGeoIpCcChk);
#endif

	gethostname(gHostnameBuf,sizeof(gHostnameBuf)-1);

	// for silly hosts that don't fully quallify their hostname
	if(strchr(gHostnameBuf,'.') == NULL)
	{
		// append the default domain name from the resolver library
		res_init();
		strncat(gHostnameBuf,".",(sizeof(gHostnameBuf)-1)-strlen(gHostnameBuf));
		strncat(gHostnameBuf,_res.defdname,(sizeof(gHostnameBuf)-1)-strlen(gHostnameBuf));
	}
	printf("Hostname: '%s'\n",gHostname);


	// if not doing attachment checks, don't do body processing
	if(!gMsExtChk)
		mlfi.xxfi_body = NULL;

	c = 0;
	smfi_setconn(gMlficonn);
	if (smfi_register(mlfi) == MI_FAILURE)
	{
		fprintf(stderr, "smfi_register failed\n");
		exit(EX_UNAVAILABLE);
	}
	else
		c = main_watcher(!gDebug & !gForeground, "/tmp/spamilter.pid");

	return c;
}
